File: app/src/main/java/com/karla/learningverbs/repository/userrepository/UserRepository.java
package com.karla.learningverbs.repository.userrepository;

import android.net.Uri;

import androidx.annotation.NonNull;

import com.google.android.gms.tasks.Continuation;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.UserProfileChangeRequest;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;
import com.karla.learningverbs.utils.LearningApplication;
import com.karla.learningverbs.utils.Tools;
import com.karla.learningverbs.utils.firebase.CustomOnCompleteListener;

public class UserRepository {
    private FirebaseAuth firebaseAuth;
    private static UserRepository instance;

    public UserRepository() {
        this.firebaseAuth = FirebaseAuth.getInstance();
    }

    public static UserRepository getInstance() {
        if (instance == null) instance = new UserRepository();
        return instance;
    }

    public boolean isLogin() {
        return firebaseAuth.getCurrentUser() != null;
    }

    public String getUserId() {
        return (firebaseAuth.getCurrentUser() != null) ? firebaseAuth.getCurrentUser().getUid() : "";

    }

    public Uri getUserUrlImage() {
        return (firebaseAuth.getCurrentUser() != null) ? firebaseAuth.getCurrentUser().getPhotoUrl() : null;
    }

    public void login(CustomOnCompleteListener<AuthResult> onCompleteListener, String email, String password) {
        FirebaseAuth.getInstance().signInWithEmailAndPassword(email, password).addOnCompleteListener(onCompleteListener);
    }

    public void singUp(CustomOnCompleteListener<AuthResult> onCompleteListener, String email, String password) {
        FirebaseAuth.getInstance().createUserWithEmailAndPassword(email, password).addOnCompleteListener(onCompleteListener);
    }

    public void recoverPassword(String email, CustomOnCompleteListener<Void> onCompleteListener) {
        FirebaseAuth.getInstance().sendPasswordResetEmail(email).addOnCompleteListener(onCompleteListener);
    }

    public void updateProfile(String userId, String urlImage, CustomOnCompleteListener<Uri> postListener) {
        FirebaseStorage storage = FirebaseStorage.getInstance();
        final StorageReference mPostReference = storage.getReference().child(LearningApplication.getInstance().getApplicationName()).child(userId + ".jpg");
        byte[] data = Tools.getImage(LearningApplication.getMyApplicationContext(), urlImage, 150, 150);
        if (data != null) {
            UploadTask uploadTask = mPostReference.putBytes(data);
            uploadTask.continueWithTask(new Continuation<UploadTask.TaskSnapshot, Task<Uri>>() {
                @Override
                public Task<Uri> then(@NonNull Task<UploadTask.TaskSnapshot> task) throws Exception {
                    if (!task.isSuccessful() && task.getException() != null) {
                        throw task.getException();
                    }
                    return mPostReference.getDownloadUrl();
                }
            }).addOnCompleteListener(postListener);
        }
    }

    public void uploadProfile(String url, CustomOnCompleteListener onCompleteListener) {
        FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
        UserProfileChangeRequest profileUpdates = new UserProfileChangeRequest.Builder()
                .setPhotoUri(Uri.parse(url))
                .build();
        user.updateProfile(profileUpdates).addOnCompleteListener(onCompleteListener);
    }

}

File: app/src/main/java/com/karla/learningverbs/kotlin/repository/UserRepository.kt
package com.karla.learningverbs.kotlin.repository

import android.net.Uri
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.UserProfileChangeRequest
import com.google.firebase.storage.FirebaseStorage
import com.karla.learningverbs.kotlin.utils.firebase.CustomOnCompleteListener
import com.karla.learningverbs.utils.LearningApplication
import com.karla.learningverbs.utils.Tools

class UserRepository {

    private val firebaseAuth: FirebaseAuth = FirebaseAuth.getInstance()


    fun isLogin(): Boolean {
        return firebaseAuth.currentUser != null
    }

    fun getUserId(): String? {
        return if (firebaseAuth.currentUser != null) firebaseAuth.currentUser!!.uid else ""
    }


    fun getUserUrlImage(): Uri? {
        return if (firebaseAuth.currentUser != null) firebaseAuth.currentUser!!.photoUrl else null
    }

    fun login(
        onCompleteListener: CustomOnCompleteListener<AuthResult?>,
        email: String?,
        password: String?
    ) {
        FirebaseAuth.getInstance().signInWithEmailAndPassword(email!!, password!!)
            .addOnCompleteListener(
                onCompleteListener
            )
    }

    fun singUp(
        onCompleteListener: CustomOnCompleteListener<AuthResult?>,
        email: String?,
        password: String?
    ) {
        FirebaseAuth.getInstance().createUserWithEmailAndPassword(email!!, password!!)
            .addOnCompleteListener(
                onCompleteListener
            )
    }

    fun recoverPassword(email: String?, onCompleteListener: CustomOnCompleteListener<Void?>) {
        FirebaseAuth.getInstance().sendPasswordResetEmail(email!!).addOnCompleteListener(
            onCompleteListener
        )
    }


    fun updateProfile(
        userId: String,
        urlImage: String,
        postListener: CustomOnCompleteListener<Uri>
    ) {
        val storage = FirebaseStorage.getInstance()
        val mPostReference =
            storage.reference.child(LearningApplication.getInstance().applicationName).child(
                "$userId.jpg"
            )
        val data = Tools.getImage(LearningApplication.getMyApplicationContext(), urlImage, 150, 150)
        if (data != null) {
            val uploadTask = mPostReference.putBytes(data)
            uploadTask.continueWithTask { task ->
                if (!task.isSuccessful && task.exception != null) {
                    throw task.exception!!
                }
                mPostReference.downloadUrl
            }.addOnCompleteListener(postListener)
        }
    }

    fun uploadProfile(url: String?, onCompleteListener: CustomOnCompleteListener<Void>) {
        val user = FirebaseAuth.getInstance().currentUser
        val profileUpdates = UserProfileChangeRequest.Builder()
            .setPhotoUri(Uri.parse(url))
            .build()
        user!!.updateProfile(profileUpdates).addOnCompleteListener(onCompleteListener)
    }

}